16 cli redesign

routesix opened the pull request:
delirium can now be used from the command line with 'dns' as a sub-command and auto-generated help.

routesix committed:
Replaced argparse' with Click

Redesigned command line interface with Click instead of argparse.

routesix committed:
Fixed Backend

Click was implemented in the setup.py file and the fakednsserver call was corrected.

routesix committed:
Update apps.py

b0urb0n commented on delirium/apps.py:
```diff
@@ -1,32 +1,34 @@
+@click.option('--listen', default=DEFAULT_LISTEN_ADDR, help='Address to listen on')
+@click.option('--port', default=DEFAULT_LISTEN_PORT, type=int, help='Port to listen on (UDP)')
+@click.option('--time', default=DEFAULT_CACHE_DURATION, type=int, help='Seconds for which cache entries should exist')
+@click.option('--range', default=DEFAULT_ADDR_RANGE, help='Range of IP addresses to randomly generate')
```
We need to use a better word than `range`.  `range` is a reserved word in Python (the `range()` function).  How about `addrs`

routesix replied:
How about --addr-pool to align with DHCP terminology?

b0urb0n commented on delirium/apps.py:
```diff
@@ -1,32 +1,34 @@
+@click.option('--port', default=DEFAULT_LISTEN_PORT, type=int, help='Port to listen on (UDP)')
+@click.option('--time', default=DEFAULT_CACHE_DURATION, type=int, help='Seconds for which cache entries should exist')
+@click.option('--range', default=DEFAULT_ADDR_RANGE, help='Range of IP addresses to randomly generate')
+@click.option('--cache_type', default=CACHE_TYPE.DICTIONARY, help='Cache type.')
```
Let's drop this option in order to facilitate https://github.com/disruptive-solutions/delirium/issues/24

b0urb0n commented on delirium/apps.py:
```diff
@@ -1,32 +1,34 @@
+
 from .const import *
 from .dns import FakeDNSServer
 
```
Two empty lines should separate imports from code

b0urb0n commented on delirium/apps.py:
```diff
@@ -1,32 +1,34 @@
+@click.group()
+def delirium():
+    pass
 
```
Two empty lines should separate all 1st level functions and classes:
```python
def something():
    pass


def something_elses():
    pass
```

b0urb0n commented:
Other than my comments above, this is a really good first attempt at this issue.  Most of my comments are related to styling (https://www.python.org/dev/peps/pep-0008/) and a requirement that you weren't yet aware of.

routesix committed:
Database Support+

Made style changes as well as implemented database-only options. Opted to use a database resident in memory as the default database location. Updated test file to accomodate a non-filepath db location.

b0urb0n commented on delirium/apps.py:
```diff
@@ -3,32 +3,32 @@
+@click.option('-a','--addr-pool', default=DEFAULT_ADDR_RANGE, help='Range of IP addresses to randomly generate')
+@click.option('-d','--db-path', default=DEFAULT_DB_PATH ,help='Path to sqlite3 database')
+def dns(laddr, lport, time, addr_pool, db_path):
+    cache_type = CACHE_TYPE.DATABASE
```
Instead of storing this constant as a variable, go ahead and just use the constant directly in the instantiation of the `FakeDNSServer` object: `s = FakeDNSServer(laddr, lport, time, addr_pool, CACHE_TYPE.DATABASE, db_path)`

b0urb0n commented on test/test_models.py:
```diff
@@ -94,7 +94,8 @@ def test_cache_range(self):
 
     def tearDown(self):
-        os.remove(DEFAULT_DB_PATH)
+        if DEFAULT_DB_PATH != ':memory:':
```
`DEFAULT_DB_PATH` is a constant and therefor will always be `":memory:"`.  You'll have to find a different way to determine if the DB was a real file.

b0urb0n replied:
I don't think it's currently possible to figure this out without significant rewiring.  Go ahead and assume it's always an in-memory database.  We will be able to solve this correctly after we switch to `pytest`.

routesix replied:
This workaround was meant to help wrap up this issue. The rewiring was going to be worked as part of #24.

b0urb0n commented on delirium/dns/fakednsserver.py:
```diff
@@ -12,7 +12,7 @@ def __init__(self, addr=DEFAULT_LISTEN_ADDR, port=DEFAULT_LISTEN_PORT, duration=
         self._port = port
 
-        self._cache = database.create_cache(ip_range, duration, cache_type, cache_path)
+        self._cache = database.create_cache(ip_range, duration, cache_path)
```
There's no longer a need for the `database.create_cache` helper function, go ahead and create a `CacheDatabase` object directly here

routesix replied:
This is also slated to be resolved as part of #24. The partial-implementation of db-only caching was only a side-effect of trying to eliminate the option out of the CLI during the CLI redesign.

b0urb0n commented on delirium/dns/models/database.py:
```diff
@@ -31,15 +32,11 @@ def init_db(path):
     return c
 
-def create_cache(addr_range, duration, cache_type, path=DEFAULT_DB_PATH):
+def create_cache(addr_range, duration, path=DEFAULT_DB_PATH):
```
This factory method is no longer required since there's only one type of cache now

routesix replied:
Recommend fixing as part of #24

b0urb0n commented on delirium/const.py:
```diff
@@ -2,7 +2,7 @@
 DEFAULT_ADDR_RANGE = '10.0.0.0-10.0.0.255'
 DEFAULT_CACHE_DURATION = 900  # 15 minutes
-DEFAULT_DB_PATH = 'fakedns.sqlite3'
+DEFAULT_DB_PATH = ':memory:' # creates the db in memory
```
PEP8 suggests 2 spaces between a line of code and an in-line comment:
```
DEFAULT_DB_PATH = ':memory:'  # creates the db in memory
```

routesix committed:
Spacing and Variable

Fixed comment formatting in const.py and eliminated redundant variable in apps.py

routesix committed:
Removed Teardown Test

Removed a test of the :memory: location for the db.
